In [5]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [6]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lendingDF = pd.read_csv("lending_data.csv")

# Review the DataFrame
lendingDF.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [7]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lendingDF['loan_status']

# Separate the X variable, the features
X = lendingDF.copy()
X.drop("loan_status", axis=1, inplace=True)

In [8]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [9]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [10]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [11]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split

# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [12]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logisticRModel = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using training data
logisticRModel.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [13]:
# Make a prediction using the testing data
predictedValues = logisticRModel.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [14]:
# Print the balanced_accuracy score of the model

# import the accuracy_score from sklearn
from sklearn.metrics import accuracy_score

# calculate the accuracy score
accuracyScore1 = accuracy_score(y_test, predictedValues)
print(f"Accuracy Score: {(accuracyScore1*100):.2f}%")

Accuracy Score: 99.18%


In [15]:
# Generate a confusion matrix for the model
from sklearn.metrics import confusion_matrix

confusion1 = confusion_matrix(y_test, predictedValues)
print(confusion1)

[[18663   102]
 [   56   563]]


In [27]:
# Print the classification report for the model
from sklearn.metrics import classification_report
classificationReport1 = classification_report(y_test, predictedValues, target_names=["Healthy", "High-Risk"])
print(classificationReport1)

              precision    recall  f1-score   support

     Healthy       1.00      0.99      1.00     18765
   High-Risk       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The Logistic Regression model predicts healthy loans with an accuracy of 99.18%. However, this accuracy is influenced by the data imbalance issue. The number of healthy loans (0) significantly outweighs the number of high-risk loans (1), suggesting that the model tends to predict loan statuses as healthy rather than identifying them as high-risk.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [17]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros_model = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros_model.fit_resample(X_train, y_train)

In [18]:
# Count the distinct values of the resampled labels data
y_resampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [19]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifierModel = LogisticRegression(solver='lbfgs', random_state=1)

classifierModel.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
predictions = classifierModel.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [22]:
# Print the balanced_accuracy score of the model 
accuracyScore2 = accuracy_score(y_test, predictions)
print(f"Accuracy Score: {(accuracyScore2*100):.2f}%")

Accuracy Score: 99.38%


In [23]:
# Generate a confusion matrix for the model
confusion2 = confusion_matrix(y_test, predictions)
print(confusion2)

[[18649   116]
 [    4   615]]


In [26]:
# Print the classification report for the model
classificationReport2 = classification_report(y_test, predictions, target_names=["Healthy", "High-Risk"])
print(classificationReport2)

              precision    recall  f1-score   support

     Healthy       1.00      0.99      1.00     18765
   High-Risk       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The accuracy score of the oversampled data is slightly higher compared to the accuracy score of the original data, with values of 99.38% and 99.18% respectively. This slight improvement in accuracy is likely attributable to the data set containing a significantly larger number of healthy loans (0) in comparison to high-risk loans (1). As a result, the improvement in the score is minimal.